### PR TITLE
 [4.7] CLOUDSTACK-9022: keep Destroyed volumes for sometime

### PR DESCRIPTION
**Pull request by ustcweizhou**
for now, the Destroyed volumes will be expunged in Storage cleanup thread, no matter when they are destroyed.
In Expunging vm thread, we expunge the Destroyed vms which have been destroyed at least 'expunge.delay' seconds. We add the similar configuration for volumes.

### CODE REVIEW

**Comment by NuxRo** [comment]
Excellent! I thought this feature was already implemented via expunge.delay, good to know it doesn't work as I thought it was.

File: server/src/com/cloud/configuration/Config.java [hunk 1]
```diff
@@ -672,6 +672,7 @@
             "86400",
             "The interval (in seconds) to wait before running the storage cleanup thread.",
             null),
+    StorageCleanupDelay("Advanced", StorageManager.class, Integer.class, "storage.cleanup.delay", "86400", "Determines how long (in seconds) to wait before actually expunging destroyed volumes. The default value = the default value of storage.cleanup.interval.", null),
```
Comment (koushik-das) [inline server/src/com/cloud/configuration/Config.java]: Please use the Configurable interface for configuration parameters instead. Refer to https://cwiki.apache.org/confluence/display/CLOUDSTACK/Configuration

File: setup/db/db/schema-452to460.sql [hunk 2]
```diff
@@ -420,3 +420,6 @@ INSERT IGNORE INTO `cloud`.`guest_os_hypervisor` (uuid,hypervisor_type, hypervis
 INSERT IGNORE INTO `cloud`.`guest_os_hypervisor` (uuid,hypervisor_type, hypervisor_version, guest_os_name, guest_os_id, created, is_user_defined) VALUES (UUID(),'KVM', 'default', 'CentOS 7', 246, utc_timestamp(), 0);
 
 UPDATE  `cloud`.`hypervisor_capabilities` SET  `max_data_volumes_limit` =  '32' WHERE  `hypervisor_capabilities`.`hypervisor_type` =  'KVM';
+
+INSERT IGNORE INTO `cloud`.`configuration` (`category`, `instance`, `component`, `name`, `value`, `default_value`, `description`) VALUES ('Advanced', 'DEFAULT', 'StorageManager', 'storage.cleanup.delay', '86400', '86400', 'Determines how long (in seconds) to wait before actually expunging destroyed volumes. The default value = the default value of storage.cleanup.interval.');
```
Comment (koushik-das) [inline setup/db/db/schema-452to460.sql]: This is not required when Configurable interface is used

File: server/src/com/cloud/configuration/Config.java (hunk 1 above)
Comment (ustcweizhou) [inline server/src/com/cloud/configuration/Config.java]: To make the configuration more readable, I will move all the storage-cleanup related configurations to Configurable interface.

**Comment by ustcweizhou** [comment]
@koushik-das Thanks for your comments. I pushed a new commit .

**Comment by wido** [comment]
Seems good! Great to see this :) Some tests are failing though, could you look into that?

**Comment by bhaisaab** [comment]
LGTM +1 (just code review, no testing was performed)

**Comment by remibergsma** [comment]
@ustcweizhou If you want this in 4.6.1, please quickly rebase and resolve conflict. Then ping me so I can run the integration tests. Thanks!

**Comment by ustcweizhou** [comment]
@remibergsma  rebased with latest master. 
There is a conflict with 4.6 because of commit 1a64c247adbeb326f7ddbe6e4feaf3159e9e740d

**Comment by remibergsma** [comment]
@ustcweizhou PR #1145 is merged to 4.6, please close this PR. Thanks!

**Comment by ustcweizhou** [comment]
@remibergsma
#1145 has conflict with master branch. you can merge this PR instead.

**Comment by DaanHoogland** [comment]
@ustcweizhou this PR is not needed as the code from 4.6 will be merged into master.

**Comment by ustcweizhou** [comment]
@DaanHoogland @remibergsma OK, I close this PR, You need to fix the conflict manually when you merge 4.6 to master.

**Pull request closed without merging**